### PR TITLE
Basic version for MPP serializer resolving from KType (typeOf() support)

### PR DESCRIPTION
**Pull request by sandwwraith**
– Standard collections
– No user-defined generic classes

### CODE REVIEW

File: runtime/commonMain/src/kotlinx/serialization/TypeTokenResolving.kt [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ */
+
+package kotlinx.serialization
+
+import kotlinx.serialization.internal.*
+import kotlin.reflect.*
+
+/**
+ * Reified version of [serializerByKType], provided for convenience.
+ * [KType] is obtained via [typeOf] function, consult its documentation
```
Comment (qwwdfsad) [inline runtime/commonMain/src/kotlinx/serialization/TypeTokenResolving.kt]: `consult` -> `refer to`.
Also, as this method will be much more popular, it is better to make its documentation self-contained.

File: runtime/commonMain/src/kotlinx/serialization/TypeTokenResolving.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * Keep in mind that this is a 'heavy' call, so result probably should be cached somewhere else.
+ *
+ * [typeOf] and [KType] API currently don't work on Kotlin/JS.
+ *
+ * @see typeOf
+ */
+@Suppress("UNCHECKED_CAST", "NO_REFLECTION_IN_CLASS_PATH", "UNSUPPORTED")
+@ImplicitReflectionSerializer
+public fun serializerByKType(type: KType): KSerializer<Any?> {
```
Comment (qwwdfsad) [inline runtime/commonMain/src/kotlinx/serialization/TypeTokenResolving.kt]: `fun serializer(type: KType)`?

`serializer(typeOf<Clz>())` is already self-desribing

File: runtime/commonMain/src/kotlinx/serialization/TypeTokenResolving.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        } as KClass<Any>
+        return when {
+            type.arguments.isEmpty() -> rootClass.serializer()
+            else -> {
+                val args = type.arguments
+                    .map { requireNotNull(it.type) { "Star projections are not allowed" } }
+                    .map(::serializerByKType)
+                // Array is not supported, see KT-32839
+                when (rootClass) {
+                    List::class, ArrayList::class -> ArrayListSerializer(args[0])
```
Comment (qwwdfsad) [inline runtime/commonMain/src/kotlinx/serialization/TypeTokenResolving.kt]: Do we want to support all potential list supertypes by inspecting superclasses of the given type?

File: runtime/commonTest/src/kotlinx/serialization/TypeTokenResolvingTest.kt [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        val myList = arrayListOf(1, 2, 3)
+        assertSerializedWithType<ArrayList<Int>>("[1,2,3]", myList)
+        assertSerializedWithType<List<Int>>("[1,2,3]", myList)
+    }
+
+    @Test
+    fun equalityOfProjectedLists() = run {
+        val myList = arrayListOf(1, 2, 3)
+        assertSerializedWithType<List<Int>>("[1,2,3]", myList)
+        assertSerializedWithType<MutableList<out Int>>("[1,2,3]", myList)
```
Comment (qwwdfsad) [inline runtime/commonTest/src/kotlinx/serialization/TypeTokenResolvingTest.kt]: `in` list is missing

File: runtime/commonMain/src/kotlinx/serialization/TypeTokenResolving.kt (hunk 3 above)
Comment (sandwwraith) [reply runtime/commonMain/src/kotlinx/serialization/TypeTokenResolving.kt]: What do you mean by supertypes? Any or something?

If you mean subtypes, the design decision was especially no – we provide default list serializers only for default list types. Moreover, KType does not contain `isSubtype/isSupertype` API

Comment (qwwdfsad) [reply runtime/commonMain/src/kotlinx/serialization/TypeTokenResolving.kt]: Yeah, I've meant subtypes.
Ok, works for me.

>Moreover, KType does not contain isSubtype/isSupertype API

`KClass.superclass` + casting to `KClass` should work

**Comment by sandwwraith** [comment]
Merged manually.

Next step would finalize design decision about how we are gonna retrieve serializers for user's generic classes on JS and Native. Then we can start integration with formats and finally make explicit passing of serializer optional.

**Pull request closed without merging**